File: modulo09-processamento-de-dados-e-fine-tuning-de-modelos/modulo-02-preparacao-datasets/dataset-preparation-pipeline-go/datarelevance/datarelevance_test.go
package datarelevance

import (
	"reflect"
	"sort"
	"testing"
)

func TestCandidatoComOs4CriteriosVerdadeirosEAceito(t *testing.T) {
	r := AvaliarCandidato(map[string]bool{
		"contemGroundTruth": true, "producaoReal": true, "cobreVariacao": true, "passaCompliance": true,
	})
	if !r.Aceito || len(r.CriteriosFalhos) != 0 {
		t.Fatalf("esperado aceito sem falhas, obtive %+v", r)
	}
}

func TestCandidatoSemGroundTruthERejeitado(t *testing.T) {
	r := AvaliarCandidato(map[string]bool{
		"contemGroundTruth": false, "producaoReal": true, "cobreVariacao": true, "passaCompliance": true,
	})
	if r.Aceito || !reflect.DeepEqual(r.CriteriosFalhos, []string{"contemGroundTruth"}) {
		t.Fatalf("esperado rejeitado por contemGroundTruth, obtive %+v", r)
	}
}

func TestCandidatoQueFalhaComplianceERejeitado(t *testing.T) {
	r := AvaliarCandidato(map[string]bool{
		"contemGroundTruth": true, "producaoReal": true, "cobreVariacao": true, "passaCompliance": false,
	})
	if r.Aceito || !reflect.DeepEqual(r.CriteriosFalhos, []string{"passaCompliance"}) {
		t.Fatalf("esperado rejeitado por passaCompliance, obtive %+v", r)
	}
}

func TestAplicacaoAos7CandidatosReais(t *testing.T) {
	esperadoAceito := map[string]bool{
		"orcamento-oficina": true, "boletim-ocorrencia": false, "foto-veiculo-danificado": false,
		"transcricao-ligacao": false, "recibo-medico": true, "prontuario-medico-completo": false,
		"cadastro-beneficiarios": false,
	}
	for _, c := range Candidatos {
		r := AvaliarCandidato(c.Criterios)
		if r.Aceito != esperadoAceito[c.ID] {
			t.Errorf("%s: esperado aceito=%v, obtive %v", c.ID, esperadoAceito[c.ID], r.Aceito)
		}
	}
}

func TestExatamente2Dos7CandidatosSaoAceitos(t *testing.T) {
	var aceitos []string
	for _, c := range Candidatos {
		if AvaliarCandidato(c.Criterios).Aceito {
			aceitos = append(aceitos, c.ID)
		}
	}
	sort.Strings(aceitos)
	if !reflect.DeepEqual(aceitos, []string{"orcamento-oficina", "recibo-medico"}) {
		t.Fatalf("esperado [orcamento-oficina recibo-medico], obtive %v", aceitos)
	}
}

File: modulo09-processamento-de-dados-e-fine-tuning-de-modelos/modulo-02-preparacao-datasets/dataset-preparation-pipeline-java/src/test/java/com/datasetprep/DataRelevanceTest.java
package com.datasetprep;

import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;

class DataRelevanceTest {

    @Test
    void candidatoComOs4CriteriosVerdadeirosEAceito() {
        var r = DataRelevance.avaliarCandidato(Map.of(
                "contemGroundTruth", true, "producaoReal", true, "cobreVariacao", true, "passaCompliance", true));
        assertThat(r.aceito()).isTrue();
        assertThat(r.criteriosFalhos()).isEmpty();
    }

    @Test
    void candidatoSemGroundTruthERejeitado() {
        var r = DataRelevance.avaliarCandidato(Map.of(
                "contemGroundTruth", false, "producaoReal", true, "cobreVariacao", true, "passaCompliance", true));
        assertThat(r.aceito()).isFalse();
        assertThat(r.criteriosFalhos()).containsExactly("contemGroundTruth");
    }

    @Test
    void candidatoQueFalhaComplianceERejeitadoMesmoComOsOutros3Verdadeiros() {
        var r = DataRelevance.avaliarCandidato(Map.of(
                "contemGroundTruth", true, "producaoReal", true, "cobreVariacao", true, "passaCompliance", false));
        assertThat(r.aceito()).isFalse();
        assertThat(r.criteriosFalhos()).containsExactly("passaCompliance");
    }

    @Test
    void aplicacaoAos7CandidatosReaisDaAmplitudeSeguros() {
        Map<String, Boolean> esperadoAceito = Map.of(
                "orcamento-oficina", true,
                "boletim-ocorrencia", false,
                "foto-veiculo-danificado", false,
                "transcricao-ligacao", false,
                "recibo-medico", true,
                "prontuario-medico-completo", false,
                "cadastro-beneficiarios", false);

        for (var candidato : DataRelevance.CANDIDATOS) {
            var resultado = DataRelevance.avaliarCandidato(candidato.criterios());
            assertThat(resultado.aceito())
                    .as("candidato %s", candidato.id())
                    .isEqualTo(esperadoAceito.get(candidato.id()));
        }
    }

    @Test
    void exatamente2Dos7CandidatosSaoAceitos() {
        List<String> aceitos = DataRelevance.CANDIDATOS.stream()
                .filter(c -> DataRelevance.avaliarCandidato(c.criterios()).aceito())
                .map(DataRelevance.Candidato::id)
                .sorted()
                .toList();
        assertThat(aceitos).containsExactly("orcamento-oficina", "recibo-medico");
    }
}
